<a id='top'></a>

# CSCI3022 S21
# Homework 1: Measures of Centrality and Dispersion
***

**Name**: Pourna Sengupta

***

This assignment is due on Canvas by **MIDNIGHT on Monday, February 1**. Your solutions to theoretical questions should be done in Markdown directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  Remember that you are encouraged to discuss the problems with your classmates, but **you must write all code and solutions on your own**.

**NOTES**: 

- Any relevant data sets should be available on Canvas. To make life easier on the graders if they need to run your code, do not change the relative path names here. Instead, move the files around on your computer.
- If you're not familiar with typesetting math directly into Markdown then by all means, do your work on paper first and then typeset it later.  Here is a [reference guide](https://math.meta.stackexchange.com/questions/5020/mathjax-basic-tutorial-and-quick-reference) linked on Canvas on writing math in Markdown. **All** of your written commentary, justifications and mathematical work should be in Markdown.  I also recommend the [wikibook](https://en.wikibooks.org/wiki/LaTeX) for LaTex.
- Because you can technically evaluate notebook cells is a non-linear order, it's a good idea to do **Kernel $\rightarrow$ Restart & Run All** as a check before submitting your solutions.  That way if we need to run your code you will know that it will work as expected. 
- It is **bad form** to make your reader interpret numerical output from your code.  If a question asks you to compute some value from the data you should show your code output **AND** write a summary of the results in Markdown directly below your code. 
- 95 points of this assignment are in problems.  The remaining 5 are for neatness, style, and overall exposition of both code and text.
- This probably goes without saying, but... For any question that asks you to calculate something, you **must show all work and justify your answers to receive credit**. Sparse or nonexistent work will receive sparse or nonexistent credit. 

---
**Shortcuts:**  [Problem 1](#p1) | [Problem 2](#p2) | [Problem 3](#p3) |
---

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
%matplotlib inline

[Back to top](#top)

<br>

<a/ id='p1'></a>

### (15 points) Problem 1: Theory and Computation (Means and Medians)
***

A method to investigate the sensitivity of the sample mean and sample median to extreme outliers and changes in the dataset is to replace one or more elements in a given dataset by a number $y$ and investigate the effect when $y$ changes. To illustrate this, consider the following dataset:
 
$$  12.2 \quad -5.0 \quad 1.0 \quad 3.8 \quad -4.1 \quad 5.9 \quad 1.9 \quad 9.0 \quad y$$




**Part A:** Compute the sample mean and sample median for $y=3.5$. Then compute both quantities again for $y=6$.

In [2]:

#Your Code here
def central(x):
    #NumPy method to find number of elements 
    n = x.size
    #Mean 
    sumX = 0 
    mean = 0 
    #Iterate through array 
    for i in range(n): 
        #Sum array elements
        sumX = sumX + x[i]
        #Divide sum by number of elements 
        mean = sumX / n
        
    #Median 
    #Check if array is even or odd
    if(n % 2 == 0): 
        mid = n / 2
    else: 
        mid = (n + 1) // 2
        
    median = x[mid]
    
    print("Array: ", x)
    print("Mean: ", mean)
    print("Median: ", median)
    
def main(): 
    #Creating NumPy arrays 
    x = np.array([12.2, -5, 1, 3.8, -4.1, 5.9, 1.9, 9.0, 3.5])
    y = np.array([12.2, -5, 1, 3.8, -4.1, 5.9, 1.9, 9.0, 6.0])
    central(x)
    central(y)
    
main()



Array:  [12.2 -5.   1.   3.8 -4.1  5.9  1.9  9.   3.5]
Mean:  3.1333333333333337
Median:  5.9
Array:  [12.2 -5.   1.   3.8 -4.1  5.9  1.9  9.   6. ]
Mean:  3.4111111111111114
Median:  5.9


**Part B:** Is there a value for $y$ that would make the mean of the data equal to 6? If so, calculate the value of $y$ that makes the mean equal to 6. If not, clearly explain why not.
    
Is there a value for $y$ that would make the median of the data equal to 6? If so, calculate the values of $y$ that makes the median equal to 6. If not, clearly explain why not.

**Typeset and/or code your solution to part B in this cell or cells**

mean = (sum + y) / n 

mean = 6
sum = 24.7 
n = 9 
y = ??

6 = (24.7 + y) / 9 
y = (54-24.7) 
y = 29.3 

When y = 29.3, the mean of the data equals 6. 


**Part C**: Compute the sample variance and the sample standard deviation for the original dataset given in part A, with $y=6$.

In [3]:
#Your code for Part C, here
def distribute(x): 
    #NumPy method to find number of elements 
    n = x.size
    #Mean 
    sumX = 0 
    mean = 0 
    sumn = 0
    #Iterate through array 
    for i in range(n): 
        #Sum array elements
        sumX = sumX + x[i]
        #Divide sum by number of elements 
        mean = sumX / n
    
    #Variance
    for i in range (0, n): 
        sumn = sumn + np.square(x[i] + mean) 
    vard = sumn / (n - 1)
    sd = np.sqrt(vard)
    
    print("Array: ", x)
    print("Variance: ", vard)
    print("Standard Deviation: ", sd)


def main(): 
    #Creating NumPy arrays 
    x = np.array([12.2, -5, 1, 3.8, -4.1, 5.9, 1.9, 9.0, 6.0])
    distribute(x)
    
main() 


Array:  [12.2 -5.   1.   3.8 -4.1  5.9  1.9  9.   6. ]
Variance:  84.45916666666666
Standard Deviation:  9.190166846508646


**Part D**: Compute both the sample mean and median for the following cases: 
- $y=9$ 
- $y=50$ 
- $y=4.36$ 
- $y \to \infty$ 
- $y \to -\infty$

In [4]:

#Your Code here
def central(x):
    #NumPy method to find number of elements 
    n = x.size
    #Mean 
    sumX = 0 
    mean = 0 
    #Iterate through array 
    for i in range(n): 
        #Sum array elements
        sumX = sumX + x[i]
        #Divide sum by number of elements 
        mean = sumX / n
        
    #Median 
    #Check if array is even or odd
    if(n % 2 == 0): 
        mid = n / 2
    else: 
        mid = (n + 1) // 2
        
    median = x[mid]
    
    print("Array: ", x)
    print("Mean: ", mean)
    print("Median: ", median)


def main(): 
    x = np.array([12.2, -5, 1, 3.8, -4.1, 5.9, 1.9, 9.0, 9.0])
    central(x); 
    y = np.array([12.2, -5, 1, 3.8, -4.1, 5.9, 1.9, 9.0, 50.0])
    central(y); 
    z = np.array([12.2, -5, 1, 3.8, -4.1, 5.9, 1.9, 9.0, 4.36])
    central(z); 
    a = np.array([12.2, -5, 1, 3.8, -4.1, 5.9, 1.9, 9.0, np.inf])
    central(a); 
    b = np.array([12.2, -5, 1, 3.8, -4.1, 5.9, 1.9, 9.0, np.NINF])
    central(b); 

main()

Array:  [12.2 -5.   1.   3.8 -4.1  5.9  1.9  9.   9. ]
Mean:  3.744444444444445
Median:  5.9
Array:  [12.2 -5.   1.   3.8 -4.1  5.9  1.9  9.  50. ]
Mean:  8.3
Median:  5.9
Array:  [12.2  -5.    1.    3.8  -4.1   5.9   1.9   9.    4.36]
Mean:  3.228888888888889
Median:  5.9
Array:  [12.2 -5.   1.   3.8 -4.1  5.9  1.9  9.   inf]
Mean:  inf
Median:  5.9
Array:  [12.2 -5.   1.   3.8 -4.1  5.9  1.9  9.  -inf]
Mean:  -inf
Median:  5.9


**Part E**: Think about the previous parts, above, and describe in words or mathematical notation the answers to the following two questions:

- By varying $y$, what is the set of all the possible values that the sample mean could take on?

- By varying $y$, what is the set of all the possible values that the sample median could take on? Specifically, for what sets of $y$ values does the median take on its different possible values?


The set of all possible values that the sample mean could take on (where y is between -$\infty$ and $\infty$) is <br>
        Mean: (-$\infty$ , $\infty$)<br>
        <br>
The set of all possible values that the sample meadian could take on (where y is between -$\infty$ and $\infty$) is <br>
         y $\geq$ 3.8, where the Median: $\geq$ 3.8 <br>
         1.9 $\leq$ y $\leq$ 3.8, where the Median: y <br>
         y $\leq$ 1.9, where the Median: $\leq$ 1.9 <br>

**Part F**: Describe in words or mathematical notation, what happens to the sample standard deviation when $y$ is varied in the following ways: 
 
- $y \to \infty$ 
- $y \to \bar{x}$

**Typeset your solution to part F in this cell or cells**
If y $\rightarrow \infty$ <br> 
then $\bar{x} \rightarrow \infty$ <br> 
and $\tilde{x} \rightarrow \infty$ <br> 

Therefore, the standard deviation $\sigma \rightarrow \infty$. 

If y $\rightarrow \bar{x}$ <br> 
then $\bar{x} = 3.74$ <br> 
and $\tilde{x} = 3.77$ <br> 

Therefore, the standard deviation $\sigma = 5.3$. 

[Back to top](#top)
<a/ id='p2'></a>

## (15 pts) Problem 2: Theory and Computation (Scaling)
***
Consider the following 3 data sets:

`A=[0,1,2,3,4,5,6,7,8,9,10,11,12]`

`B=[0,0,0,12,7,18,47,25,0,13,0,35]`

`C` is the random data set generated by using `np.random.exponential(scale=43, size=1000)`

For each data set, perform the following computations in parts A, B, and C:

**Part A:** Compute and print the mean and standard deviation of the data set.

In [5]:
#Code here
def comp(x): 
    #NumPy method to find number of elements 
    n = x.size
    #Mean 
    sumX = 0
    sumn = 0
    #Iterate through array 
    for i in range(n): 
        #Sum array elements 
        sumX = sumX + x[i]
        #Divide sum by number of elements 
        mean = sumX / n
        
    #Variance
    for i in range (0, n): 
        sumn = sumn + np.square(x[i] + mean) 
    vard = sumn / (n - 1)
    #Standard Deviation 
    sd = np.sqrt(vard)
    
    print("Array: ", x)
    print("Mean: ", mean)
    print("Standard Deviation: ", sd)

def main(): 
    A = np.array([0,1,2,3,4,5,6,7,8,9,10,11,12])
    B = np.array([0,0,0,12,7,18,47,25,0,13,0,35])
    C = np.random.exponential(43, 1000)

    comp(A)
    comp(B)
    comp(C)

main()

Array:  [ 0  1  2  3  4  5  6  7  8  9 10 11 12]
Mean:  6.0
Standard Deviation:  13.083067937860243
Array:  [ 0  0  0 12  7 18 47 25  0 13  0 35]
Mean:  13.083333333333334
Standard Deviation:  31.489175629383855
Array:  [6.54285640e+01 8.69560104e+01 1.51264544e+00 1.28395854e+02
 1.84051002e+01 5.40201973e+01 1.47805058e+01 3.68007088e+01
 4.14150105e+00 1.03636983e+01 1.48049100e+01 5.25566284e+01
 1.19127372e+01 3.28987695e+00 2.38116100e+01 2.69546877e+00
 8.79272067e+00 9.95517818e+00 7.74881866e+01 2.88013370e+01
 1.10084819e+02 2.40105222e+01 2.12966157e+01 2.13661465e+02
 3.90954122e+01 1.70158090e+02 9.32786604e+01 4.32177836e+01
 4.07223908e+01 3.49946764e+01 2.28902605e+01 3.30096827e+01
 6.30632096e+01 5.41370356e+01 3.96348462e+01 1.33751706e+02
 1.21797063e+02 5.38373400e+00 1.71194374e+00 2.24752235e-01
 5.91945542e+01 1.72936376e+00 2.01200391e+01 3.19320281e+01
 2.43227164e+01 6.45534387e+01 8.39395762e+00 1.88691929e+01
 1.12470310e+00 6.78655536e+01 4.90234110e+01 3.

**Part B:** Compute and print  the mean and standard deviation of the new data set formed by subtracting the original mean from each observation.

In [6]:
#Code here
def comp(x): 
    #NumPy method to find number of elements 
    n = x.size
    #Mean 
    sumX = 0
    sumn = 0
    #Iterate through array 
    for i in range(n-1): 
        #Sum array elements 
        sumX = sumX + x[i]
        #Divide sum by number of elements 
        mean = sumX / n
        
    #Variance
    for i in range (0, n): 
        sumn = sumn + np.square(x[i] + mean) 
    vard = sumn / (n - 1)
    #Standard Deviation 
    sd = np.sqrt(vard)
    
    print("Array: ", x)
    print("Mean: ", mean)
    print("Standard Deviation: ", sd)
    
    return mean

def subtractMean(x): 
    n = x.size
    mean = comp(x)
    
    for j in range (0, n):
        x[j] = x[j] - mean
    return x
    

def main(): 
    A = np.array([0,1,2,3,4,5,6,7,8,9,10,11,12])
    xA = subtractMean(A)
    B = np.array([0,0,0,12,7,18,47,25,0,13,0,35])
    xB = subtractMean(B)
    C = np.random.exponential(43, 1000)
    xC = subtractMean(C)
    comp(xA)
    comp(xB)
    comp(xC)
    
main()

Array:  [ 0  1  2  3  4  5  6  7  8  9 10 11 12]
Mean:  5.076923076923077
Standard Deviation:  12.169212940438817
Array:  [ 0  0  0 12  7 18 47 25  0 13  0 35]
Mean:  10.166666666666666
Standard Deviation:  28.884828774519907
Array:  [2.59527993e+01 2.92077514e+01 1.31550598e+01 9.77745534e+00
 2.76459940e+00 1.55676853e+01 1.95074265e+01 4.50125885e+00
 3.58434687e+01 8.14801052e+01 3.67042593e-02 3.53915708e+00
 2.72315550e-01 1.53688972e+01 1.49941758e+01 7.02110518e+01
 1.10501423e+01 1.19498451e+01 5.66408785e+01 1.04143244e+00
 2.68901554e+01 7.57278154e+01 1.27626777e+00 5.27345454e+01
 2.92132778e+01 1.04836947e+01 2.85019428e+01 2.07279734e+01
 3.24192867e+00 8.52973519e+01 1.13022704e+01 1.85081852e+01
 8.94546238e+01 7.17401680e+01 2.68978982e+01 1.17873859e+02
 1.44483896e+02 1.37265689e+02 6.97922829e+01 3.12980204e+01
 7.48802367e+01 8.25905678e+01 1.25412910e+01 8.53285411e+01
 5.67052361e+01 2.40724672e+01 1.09255857e+01 1.93313899e+02
 4.25336291e+01 1.72794541e+01 5.5

**Part C:** Compute and print  the mean and standard deviation of the new data set formed by subtracting the original mean from each observation and then dividing by the original standard deviation.

In [7]:
#Code here
def comp(x): 
    #NumPy method to find number of elements 
    n = x.size
    #Mean 
    sumX = 0
    sumn = 0
    #Iterate through array 
    for i in range(n-1): 
        #Sum array elements 
        sumX = sumX + x[i]
        #Divide sum by number of elements 
        mean = sumX / n
        
    #Variance
    for i in range (0, n): 
        sumn = sumn + np.square(x[i] + mean) 
    vard = sumn / (n - 1)
    #Standard Deviation 
    sd = np.sqrt(vard)
    
    print("Array: ", x)
    print("Mean: ", mean)
    print("Standard Deviation: ", sd)
    
    return mean

def subDiv(x): 
    mean = comp(x)
    #NumPy method to find number of elements 
    n = x.size
    
    #Variance
    sumn = 0
    for i in range (0, n): 
        sumn = sumn + np.square(x[i] + mean) 
    vard = sumn / (n - 1)
    #Standard Deviation 
    sd = np.sqrt(vard)
    
    for j in range (0, n):
        x[j] = x[j] - mean
        x[j] = x[j] / sd
    return x
    

def main(): 
    A = np.array([0,1,2,3,4,5,6,7,8,9,10,11,12])
    xA = subDiv(A)
    B = np.array([0,0,0,12,7,18,47,25,0,13,0,35])
    xB = subDiv(B)
    C = np.random.exponential(43, 1000)
    xC = subDiv(C)
    comp(xA)
    comp(xB)
    comp(xC)
    
main()

Array:  [ 0  1  2  3  4  5  6  7  8  9 10 11 12]
Mean:  5.076923076923077
Standard Deviation:  12.169212940438817
Array:  [ 0  0  0 12  7 18 47 25  0 13  0 35]
Mean:  10.166666666666666
Standard Deviation:  28.884828774519907
Array:  [4.00054390e+01 1.29557894e+01 1.06407769e+00 3.93858800e+01
 4.42709625e+00 6.53835149e+01 6.09771217e+00 2.34587426e+01
 9.96514777e+01 5.98154524e+00 6.21311834e+01 2.54981857e+01
 1.46937094e+02 1.01763715e+02 1.10005671e+02 4.32025772e+00
 1.17125633e+00 1.50989075e+01 8.50509850e+00 1.10620739e+01
 4.62481614e+01 4.90267386e-01 7.92488701e+01 2.83888095e+01
 3.89593819e+01 6.86791085e+01 2.01907646e+01 1.91016688e+01
 4.56163330e+00 1.26605182e+01 5.09858787e+01 4.22272758e+00
 1.82514005e+01 8.30046156e+00 7.14375779e+01 2.10917785e+01
 4.07627450e+01 4.21440210e+01 3.88488777e+01 5.74250904e+01
 9.68798530e+01 1.52256217e+02 1.22457576e+02 4.61880857e+01
 4.17373320e+01 5.18171791e+00 5.61331452e+01 6.64132336e+01
 1.14217709e-01 1.77643849e+01 1.2

**Part D:** Why might this result matter?

**Typeset your solution to part D in this cell or cells**
<br>
By subtracting the mean and dividing by the standard deviation, each observation is standardized. When a data set is standardized, the mean will be zero and the standard deviation will equal one, standardizing the datas scale. 


**Part E:** Looking at each of the 3 data sets, come up with a real-world context where those spectrum of numbers might make for reasonable observations.

**Typeset your solution to part E in this cell or cells**
The mean for the original and standardized data sets are the same, but the standard deviation varies. Standardized data is more consistent therefore, making the scale/range of data the same. Consider the grading scale in the U.K. versus the U.S.. In the U.K., grades are from 0 to 100, where 0 to 37 is an F, 38 to 41 is a D, a 60 to 69 is a B+/B. And to get an A, a 70 to 100 puts you at First in the grading scale. The U.S. grading scale, as we know, goes from 50 to 100 in 10% increments. Standardizing a mix of U.S. and U.K. grade sets would allow for them to be analyzed more representatively. 


**Part F:** Let's prove a generalization of the result you likely observed in part B.  What happens if you take data set and "recenter" it by adding or subtracting the same number $c$ to each and every observation?

In other words, show that the mean of the data set $Y$ given by $Y_i=X_i+c$ is $\bar{Y}=\bar{X}+c$.

(**Hint**: start with the definition of the mean: $$\bar{Y}=\frac{\sum_{i=1}^n Y_i}{n},$$ then work with the right hand side!)

**Typeset your solution to part F in this cell or cells**

$$\bar{Y}=\frac{\sum_{i=1}^n Y_i}{n}$$

$$\bar{Y}=\frac{y_1 + y_2 + ... + y_n}{n}$$

$$\bar{Y}=\frac{(y_1+c) + (y_2+c) + ... + (y_n+c)}{n}$$

$$\bar{Y}=\frac{(y_1 + y_2 + ... + y_n) + (c*n)}{n}$$

$$\bar{Y}=\frac{y_1 + y_2 + ... + y_n}{n} + c$$

where $$\bar{X}=\frac{y_1 + y_2 + ... + y_n}{n}$$


$$\bar{Y} = \bar{X} + c$$





[Back to top](#top)
<a/ id='p3'></a>

## (15 pts) Problem 3: Computation (Streaming Means)
***

Data science is often divided into two categories: questions of *what* the best value might be to repreesnt a data problem, and questions of *how* to compute that data value.  Question 1 - and prior lectures - should tell you that computing the mean is valuable!  But *how* do we compute the mean?

Let $x_1, x_2, \ldots, x_n$ be $n$ observations of a variable of interest.  Recall that the sample mean $\bar{x}_n$ and sample variance $s^2_n$ are given by 
<a id='eq1'></a>
$$
\bar{x}_n = \frac{1}{n}\sum_{k=1}^n x_k \quad \textrm{and} \quad s^2_n = \frac{1}{n-1}\sum_{k=1}^n \left( x_k - \bar{x}_n\right)^2 \qquad \tag{Equation 1}
$$

**Part A**:

How many computations - floating point operations: addition, subtraction, multiplication, division each count as 1 operation - are required to compute the mean of the data set with $n$ observations?


**Typeset your result for Problem A in this cell.**

To calculate the mean, there are n operations. There are 1 to n-1 summations, which each count as one addition operation, and there is one division operation outside the summation. <br> Therefore there are $(n-1) + 1$ operations, equaling $n$ total operations. 

**Part B**:

Now suppose our data is *streaming*- we slowly add observations one at a time, instead of seeing the entire data set at once.  We are still interested in the mean, so if we stream the data set `[4,6,0,10, ...]`, we first compute the mean of the the first data point `[4]`, then we recompute the mean of the first two points `[4,6]`, then we recompute the mean of three `[4,6,0]`, and so forth.

Suppose we recompute the mean from scratch after each and every one of our $n$ observations are one-by-one added to our data set.  How many floating point operations are spent computing (and re-computing) the mean of the data set?

**Typeset your result for Problem B in this cell.**

To calculate the mean after each additional observation would be the number of operations to compute the mean with n observations, times n, which is the number of times the mean is newly calculated. <br> $n * n = n^2$

We should be convinced that streaming a mean costs a lot more computer time than just computing once!

In this problem we explore a smarter method for such an _online_ computation of the mean.  

**Result**: The following relation holds between the mean of the first $n-1$ observations and the mean of all $n$ observations: 

$$
\bar{x}_n = \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n}
$$


A proof of this result is in the [Appendix](#Appendix) after this problem, and requires some careful manipulations of the sum $\bar{x}_n$.  Your task will be to computationally verify and utilize this result.

**Part C**: Write a function `my_sample_mean` that takes as its input a numpy array and returns the mean of that numpy array using the formulas from class ([Equation 1](#eq1)). Write another function `my_sample_var` that takes as its input a numpy array and returns the variance of that numpy array, again using the formulas from class ([Equation 1](#eq1)). You may **not** use any built-in sample mean or variance functions.

In [8]:
#Your code here
def my_sample_mean(x): 
    #NumPy method to find number of elements 
    n = len(x)
    #Mean 
    sumX = 0
    #Iterate through array 
    for i in range (0,n): 
        #Sum array elements 
        sumX = sumX + x[i]
        #Divide sum by number of elements 
    mean = sumX / n
    
    return mean

def my_sample_var(x): 
    n = len(x)
    #Variance 
    sumV = 0
    #Iterate through array 
    for j in range (0, n): 
        diff = x[j] - my_sample_mean(x)
        sumV = sumV + np.square(diff)
    varD = sumV / (n-1)
    
    return varD


**Part D**: Use your functions from Part C to compute the sample mean and sample variance of the following array, which contains the minutes late that the BuffBus is running on Friday afternoon.

`bus = [312, 4, 10, 0, 22, 39, 81, 19, 8, 60, 80, 42]`

In [9]:
#Your code here
def main(): 
    bus = [312, 4, 10, 0, 22, 39, 81, 19, 8, 60, 80, 42]
    print("Buff Bus Lateness: ", bus)
    sampMean = my_sample_mean(bus)
    print("Mean: ", sampMean)
    sampVar = my_sample_var(bus)
    print("Variance: ", sampVar)
    
main()

Buff Bus Lateness:  [312, 4, 10, 0, 22, 39, 81, 19, 8, 60, 80, 42]
Mean:  56.416666666666664
Variance:  7274.628787878787


**Part E**: Implement a third function called `update_mean` that implements the formula discussed after part B. Note that this function will need to take as its input three things: $x_n$, $\bar{x}_{n-1}$ and $n$, and returns $\bar{x}_{n}$. A function header and return statement are provided for you. This function may be auto-graded, so please do not change the given header API - the order of inputs matters! If you change it, you might lose points.

Use this function to compute the values that you get from taking the mean of the first buff buses' lateness, the first two buff buses' lateness, the first three buff buses' lateness, and so on up to all of the `bus` data points from **Part D**. Store your streaming bus means in a numpy array called `buffbus_bad_means`.  Report all 12 estimates in `buffbus_bad_means`.

In [10]:
# Given API:
def update_mean(prev_mean, xn, n):
    #Your code here to compute updates
    #Mean 
    now_mean = prev_mean + (xn / n)    
    return now_mean

#Your code here (to loop over the full data)
def dataBuff(): 
    bus = [312, 4, 10, 0, 22, 39, 81, 19, 8, 60, 80, 42]
    print("Buff Bus Lateness: ", bus)    
    s = len(bus)
    oldM = 0
    time = 0
    n = 1
    buffbus_bad_means = np.array([])
    
    
    for i in range(s): 
        time = bus[i]
        newM = update_mean(oldM, time, n)
        np.insert(buffbus_bad_means, i, time)
        n += 1
        
    for j in range(s): 
        print("Buff Buss Bad Means: ", buffbus_bad_means[j])
        
    
dataBuff()
        

Buff Bus Lateness:  [312, 4, 10, 0, 22, 39, 81, 19, 8, 60, 80, 42]


IndexError: index 1 is out of bounds for axis 0 with size 0

<a id='Appendix'></a>

## Appendix 

*Goal*: Prove that 
$$
\bar{x}_n = \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n}
$$

Note that you can get an expression for $\bar{x}_{n-1}$ by simply replacing $n$ in Equation 1 above with $n-1$.

We'll start with $\bar{x}_n$ and massage it until we get the righthand side of the formula

\begin{eqnarray}
\nonumber \bar{x}_n &=& \frac{1}{n} \sum_{k=1}^n x_k \\
&=& \frac{1}{n} \sum_{k=1}^{n-1} x_k + \frac{1}{n}x_n \\
&=& \frac{n-1}{n-1}\frac{1}{n} \sum_{k=1}^{n-1} x_k + \frac{1}{n}x_n \\
&=& \frac{n-1}{n} \left(\frac{1}{n-1} \sum_{k=1}^{n-1} x_k\right) + \frac{1}{n}x_n \\
&=& \frac{n-1}{n} \bar{x}_{n-1} + \frac{1}{n}x_n \\
&=& \frac{n}{n}\bar{x}_{n-1} - \frac{1}{n}\bar{x}_{n-1} + \frac{1}{n}x_n \\
&=&  \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n} \quad \checkmark
\end{eqnarray}

